<h2 id="Metode-NumerikPendahuluan-Analisis-Error-" style="text-align: center;"><strong>Metode Numerik</strong><br /><strong><span style="color: blue;">Pendahuluan Analisis Error </span></strong><br /><img style="height: 200px;" src="images/MetNum.png" alt="" /></h2>
<h2 id="(C)-Taufik-Sutanto---2019" style="text-align: center;">(C) Taufik Sutanto - 2019</h2>
<h2 id="tau-data-Indonesia-~-https://tau-data.id" style="text-align: center;">tau-data Indonesia ~ <a href="https://tau-data.id" target="_blank"><span style="color: #0009ff;">https://tau-data.id</span></a></h2>

<p><strong><big>III. Analisa Error:</big></strong></p>

<ol>
	<li>Bilangan Binari</li>
	<li>Floating Point</li>
	<li>Error Absolut VS&nbsp;Relatif</li>
	<li>Significant digits</li>
	<li>Truncation VS Round Error</li>
	<li>Cancellation</li>
	<li>Order of Approximation &amp; Convergence</li>
	<li>Propagation Error</li>
</ol>


In [1]:
x = 0.1
type(x)

float

<p><img alt="" src="images/Binary2Decimal.png" style="width: 587px; height: 281px;" /></p>


<p><img alt="" src="images/Decimal2Binary.jpg" style="width: 500px; height: 375px;" /></p>


<p><img alt="" src="images/Fraction2Binary.jpg" style="width: 500px; height: 375px;" /></p>


<h1 id="Practice"><u>Practice</u></h1>

<h2 id="Convert-$(5.1)_{10}$-to-binary">Convert <span class="MathJax_Preview" style="color: inherit;"><span class="MJXp-math" id="MJXp-Span-64"><span class="MJXp-mo" id="MJXp-Span-65" style="margin-left: 0em; margin-right: 0em;">(</span><span class="MJXp-mn" id="MJXp-Span-66">5.1</span><span class="MJXp-msubsup" id="MJXp-Span-67"><span class="MJXp-mo" id="MJXp-Span-68" style="margin-left: 0em; margin-right: 0.05em;">)</span><span class="MJXp-mrow MJXp-script" id="MJXp-Span-69" style="vertical-align: -0.4em;"><span class="MJXp-mn" id="MJXp-Span-70">10</span></span></span></span></span><script type="math/tex" id="MathJax-Element-13">(5.1)_{10}</script> to binary</h2>


# Why learning the base of numbers important in Numerical methods?
## Let's calculate this:using computer (Python)
# $\sum_{n=1}^{10^6}0.1$
## What is the expected result?

In [2]:
from tqdm import tqdm_notebook as tqdm

In [3]:
# Here is the result according to Python
dSum = 0 
for i in tqdm(range(10**6)): # ini artinya for i dari 0 sampai (10^6)-1
    dSum = dSum + 0.1 # or "dSum += 0.1"
print(dSum)
# is the result correct? Why?


100000.00000133288


<h1>Computer can&#39;t store Real Numbers!, it store &quot;floating point&quot; number</h1>
<img alt="" src="images/floating_points.png" style="width: 517px; height: 189px;" />

<h1 id="Bilangan-floating-point-tidak-tersebar-merata-di-garis-bilangan">Bilangan floating point tidak tersebar merata di garis bilangan</h1>

<p><img alt="" src="images/Sebaran_floating_points.png" style="width: 500px; height: 329px;" /></p>

In [4]:
# System floating point di Python
import sys
sys.float_info

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)

<p><img alt="" src="images/Float_Info.PNG" style="width: 550px; height: 406px;" /></p>


In [5]:
# Testing it out
x = 1.7976931348623157e+308
print(x*2) # OverFlow
e = 2.120446049250313e-17 # underflow < e < Epsilon
print(1+e)

inf
1.0


<h1>Perhatikan:</h1>

<h1>Karena ini deret harmonik ($\sum_{i=1}^\inf 1/i$) di komputer konvergen dan penjumlahan tidak bersifat asosiatif. Why????....</h1>


<h1 id="Tapi-jika-pakai-modul-&quot;Numpy&quot;">Tapi jika pakai modul &quot;Numpy&quot;</h1>

<p><a href="https://docs.scipy.org/doc/numpy/user/basics.types.html" target="_blank">https://docs.scipy.org/doc/numpy/user/basics.types.html</a></p>

<p><img alt="" src="images/Numpy_Types.png" style="width: 500px; height: 363px;" /></p>


<p><img alt="" src="images/Scientific_Notations.png" style="width: 284px; height: 205px;" /></p>


In [8]:
import math
math.pi

3.141592653589793

<p><img alt="" src="images/Abs_Rel_Errors.png" style="width: 550px; height: 367px;" /></p>


<h1 id="Diskusi">Diskusi</h1>

<ol>
	<li>Apa perbedaan error absolut dan relatif?</li>
	<li>Bilamana error yang satu lebih cocok digunakan dibandingkan error yang lain?<br />
	Misal error absolut better than relatif atau sebaliknya.</li>
	<li>Bolehkah menghitung error dengan cara lain?&nbsp;</li>
</ol>


<h1 id="Latihan"><u>Latihan</u></h1>

<h2>Jika sin(3) dihitung dengan menggunakan deret McLaurin (5 suku), tentukan error absolute dan relatifnya. &quot;<strong>Asumsikan</strong>&quot; nilai sinus dari kalkulator atau program sebagai nilai sejatinya.&nbsp;</h2>

<p><img alt="" src="images/McLaurin_Series.jpg" style="width: 400px; height: 117px;" /></p>


In [6]:
# Dengan Python
import math
x = 3
Y_real = math.sin(x)
Y_cap = x - (x**3/math.factorial(3)) + (x**5/math.factorial(5))-(x**7/math.factorial(7))
print('Nilai eksak = ',Y_real,' *Assumed')
print('Nilai pendekatan = ',Y_cap)
print('Error Absolut = ', abs(Y_real-Y_cap))
print('Error Relatif = ', abs(Y_real-Y_cap)/abs(Y_real))
# Be critical ... apa kekurangan dari menghitunng fungsi sinus dengan deret ini?

Nilai eksak =  0.1411200080598672  *Assumed
Nilai pendekatan =  0.09107142857142847
Error Absolut =  0.050048579488438744
Error Relatif =  0.3546526121739355


<h1 id="Calculating-Significant-Digits">Calculating Significant Digits</h1>

<p>Given a relative error&nbsp;<i>E</i><sub>rel</sub>, find the largest integer&nbsp;<i>n</i>&nbsp;such that&nbsp;<i>E</i><sub>rel</sub>&nbsp;&lt; 0.5 10<sup>-n</sup></p>

<ol>
	<li> 
	What is the number of significant digits of the approximation 3.14 to the value &pi;?
	<p><i>E</i><sub>rel</sub>&nbsp;= |3.14 - &pi;|/|&pi;| &asymp; 0.00051 &le; 0.005 = 0.5 &sdot; 10<sup>-2</sup>, and therefore it is correct to two significant digits.</p>
	</li>
	<li>
	<p>What is the number of significant digits of the label 240 &Omega; when the correct value is 243.32753 &Omega;?</p>
	</li>
    <li>
	<p>E<sub>rel</sub>&nbsp;= |240 - 243.32753|/|243.32753| &asymp; 0.014 &le; 0.05 = 0.5 &sdot; 10<sup>-1</sup>, and therefore it is correct to one significant digit.</p>
	</li>
	<li>
	<p>To how many significant digits is the approximation 1.998532 when the actual value is 2.001959?</p>
        
        
 <p>E<sub>rel</sub>&nbsp;= |1.998532 - 2.001959|/|2.001959| &asymp; 0.0017 &le; 0.005 = 0.5 &sdot; 10<sup>-2</sup>&nbsp;and therefore it is correct to two digits.</p>

<h1>Latihan:</h1>

<ol>
	<li>To how many significant digits does 22/7 approximate &pi;?</li>
	<li>To how many significant digits does 355/113 approximate pi &pi;?</li>
</ol>


In [10]:
pi = math.pi
rel_error = abs(355/113 - pi)/pi
rel_error*2

1.698273575348122e-07

In [11]:
# Menentukan Significant Digits Menggunakan Python
relatif_error = rel_error #0.0036
significant_digits = int( -math.log10( relatif_error / 0.5 ) )
print(significant_digits)

6


<p><img alt="" src="images/truncation_error.png" style="width: 530px; height: 341px;" /></p>


<p><img alt="" src="images/Truncation_error_example.PNG" style="width: 550px; height: 187px;" /></p>


<h1 id="Rounding-error-memiliki-error-≤5">Rounding error memiliki error <span class="MathJax_Preview" style="color: inherit"><span class="MJXp-math" id="MJXp-Span-207"><span class="MJXp-mo" id="MJXp-Span-208" style="margin-left: 0.333em ; margin-right: 0.333em">&le;</span><span class="MJXp-mn" id="MJXp-Span-209">5</span></span></span></h1>

<p>Misal: <span class="MathJax_Preview" style="color: inherit;"><span class="MJXp-math" id="MJXp-Span-235"><span class="MJXp-mi MJXp-italic" id="MJXp-Span-236">x</span><span class="MJXp-mo" id="MJXp-Span-237" style="margin-left: 0.333em; margin-right: 0.333em;">=</span><span class="MJXp-mn" id="MJXp-Span-238">77.561</span></span></span><script type="math/tex" id="MathJax-Element-44">x = 77.561</script> maka <span class="MathJax_Preview" style="color: inherit;"><span class="MJXp-math" id="MJXp-Span-239"><span class="MJXp-mrow" id="MJXp-Span-240"><span class="MJXp-munderover" id="MJXp-Span-241"><span><span class="MJXp-over"><span style="margin-bottom: -1.17em;"><span class="MJXp-mo" id="MJXp-Span-243" style="margin-left: 0px; margin-right: 0px;">ˉ</span></span><span><span class="MJXp-mi MJXp-italic" id="MJXp-Span-242">x</span></span></span></span></span></span><span class="MJXp-mo" id="MJXp-Span-244" style="margin-left: 0.333em; margin-right: 0.333em;">=</span><span class="MJXp-mn" id="MJXp-Span-245">77.6</span></span></span><script type="math/tex" id="MathJax-Element-45">\bar x=77.6</script></p>


<p>Tapi bagaimana dengan 0.5? (misal 77.5)</p>


<p><img alt="" src="images/Rounding_half.PNG" style="width: 550px; height: 425px;" /></p>


<h1 id="Cancellation-Error">Cancellation Error</h1>

<p>(1+eps)-(1-eps)= 1-1=0</p>

<p>Padahal seharusnya hasil 2*eps masih bisa dihandel mesin, namun hasil ini hilang.</p>

<p>Contoh nyata, pada deret $e^x=1+x+x2/2!+x3/3!+ ... $</p>

<p>Untuk x&lt;0</p>

<p>Contoh lain yang terkenal adalah solusi &quot;kecap&quot; penyelesaian polinomial (jika $b^2$ ~ 4ac)</p>


<p><img alt="" src="images/other_example_Cancellation_Error.png" style="width: 550px; height: 376px;" />Contoh Lain</p>


<h1>Order of Approximation (Big-Oh):</h1>

<p><img alt="" src="images/big-oh.png" style="width: 752px; height: 96px;" /></p>
Contoh $f(h) = 2h^3 + 77$ adalah $O(h^3)$ atau $O(h^5)$, tapi bukan $O(h^2)$
Why?

<h1 id="Contoh-Aplikasi">Contoh Aplikasi</h1>

<p><img alt="" src="images/Deret_Taylor.png" style="width: 400px; height: 232px;" /></p>


<h1 id="Perhatikan-:">Perhatikan :</h1>

<p>Order of convergence maupun Deret Taylor tidak memberikan nilai eksak errornya. Hanya perkiraan &quot;order&quot;/besaran error akibat penggunaan m jumlah suku dari deretnya. Dengan kata lain nilai c cuma dipastikan ada pada suatu interval, tapi tidak tau nilai pastinya.</p>

<p>&nbsp;</p>

<p>Lihat contoh lain di buku atau sumber lain di internet.</p>


## Order of Convergence
<p><img alt="" src="images/Order_of_Convergence.PNG" style="width: 600px; height: 290px;" /></p>


# Contoh Lain?

Video Source: https://www.youtube.com/watch?v=8IfiTOhvnPM

<video controls src="images/001 Errors and error propagation - YouTube.MP4" />

<h1>End of Week III</h1>

<hr />
<p>&nbsp;</p>


In [12]:
from collections import Counter as count

A = [(1,2), (1,2), (3,2), (3,2), (3,2), (4,9)]
count(A)

Counter({(1, 2): 2, (3, 2): 3, (4, 9): 1})